File: rest-service/src/main/java/mb/lib/util/Hash.java
package mb.lib.util;

import java.nio.charset.StandardCharsets;

public class Hash
{
  private static final int HashSize = 16;

  public static String hashToString(byte[] val) {
    if (val.length != HashSize)
      throw new IllegalArgumentException("Invalid hash value length.");

    var out = new byte[val.length * 2];

    for (var i = 0; i < val.length; i++) {
      byteToHex(out, i*2, val[i]);
    }

    return new String(out);
  }

  public static byte[] stringToHash(String val) {
    var len = val.length();

    if (len != HashSize * 2)
      throw new IllegalArgumentException("Invalid hash value length.");

    var out = new byte[len / 2];
    var in  = val.getBytes(StandardCharsets.UTF_8);

    for (var i = 0; i < len; ) {
      out[i/2] = charsToByte(in[i++], in[i++]);
    }

    return out;
  }

  static void byteToHex(byte[] buf, int pos, byte b) {
    buf[pos]   = byteToChar((byte) ((b >> 4) & 0x0F));
    buf[pos+1] = byteToChar((byte) (b & 0x0F));
  }

  static byte byteToChar(byte b) {
    if (b > 15)
      throw new IllegalArgumentException("Cannot translate values > 15 to a hex digit.");

    return (byte) (b > 9 ? 'A' + (b - 10) : '0' + b);
  }

  static byte charsToByte(byte a, byte b) {
    return (byte) (charToByte(a) << 4 | charToByte(b));
  }

  static byte charToByte(byte v) {
    if (v >= 'a' && v <= 'f')
      return (byte) (v - 'a' + 10);
    if (v >= 'A' && v <= 'F')
      return (byte) (v - 'A' + 10);
    if (v >= '0' && v <= '9')
      return (byte) (v - '0');

    throw new IllegalArgumentException("Invalid hex character " + (char) v);
  }

}

File: rest-service/src/main/java/mb/lib/util/hash.kt
package mb.lib.util

private const val HashSize = 16

fun hashToString(value: ByteArray): String {
  if (value.size != HashSize)
    throw IllegalArgumentException("Invalid hash value length.")

  val out = ByteArray(value.size * 2)

  for (i in value.indices) {
    byteToHex(out, i*2, value[i])
  }

  return String(out)
}

fun stringToHash(value: String): ByteArray {
  val len = value.length

  if (len != HashSize * 2)
    throw IllegalArgumentException("Invalid hash value length.")

  val out = ByteArray(len/2)
  val inp = value.toCharArray()

  var i = 0
  while (i < len) {
    out[i/2] = charsToByte(inp[i++], inp[i++])
  }

  return out
}

fun byteToHex(buf: ByteArray, pos: Int, b: Byte) {
  buf[pos]   = byteToChar(((b.toInt() shr 4) and 0x0F).toByte())
  buf[pos+1] = byteToChar((b.toInt() and 0x0F).toByte())
}

fun byteToChar(b: Byte): Byte {
  return if (b > 15)
    throw IllegalArgumentException("Cannot translate values > 15 to a hex digit.")
  else
    (if (b > 9) 'A' + (b - 10) else '0' + b.toInt()).code.toByte()
}

fun charsToByte(a: Char, b: Char) = (charToByte(a) shl 4 or charToByte(b)).toByte()

fun charToByte(v: Char) = when(v.code) {
  in 97..102 -> v - 97 + 10
  in 65..70  -> v - 65 + 10
  in 48..57  -> v - 48
  else         -> throw IllegalArgumentException("Invalid hex character $v")
}.code
